populate.go

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Populate sets targets with values from the dependency injection container
+// before application start. All targets must be pointers to the values
+// that must be populated. Pointers to structs that embed In are supported,
+// which can be used to populate multiple values in a struct.
+func Populate(targets ...interface{}) Option {
+	// Validate all targets are non-nil pointers.
+	targetTypes := make([]reflect.Type, len(targets))
+	for i, t := range targets {
+		if t == nil {
+			return invokeErr(fmt.Errorf("failed to Populate: target %v is nil", i+1))
+		}
+		rt := reflect.TypeOf(t)
+		if rt.Kind() != reflect.Ptr {
+			return invokeErr(fmt.Errorf("failed to Populate: target %v is not a pointer type, got %T", i+1, t))
+		}
+
+		targetTypes[i] = reflect.TypeOf(t).Elem()
+	}
+
+	// Build a function that looks like:
+	//
+	// func(t1 T1, t2 T2, ...) {
+	//   *targets[0] = t1
+	//   *targets[1] = t2
+	//   [...]
+	// }
+	//
+	fnType := reflect.FuncOf(targetTypes, nil, false /* variadic */)
+	fn := reflect.MakeFunc(fnType, func(args []reflect.Value) []reflect.Value {
+		for i, arg := range args {
+			reflect.ValueOf(targets[i]).Elem().Set(arg)
+		}
+		return nil
```

Comment:
"return nil" is not obvious here.  Doc says https://golang.org/pkg/reflect/#MakeFunc

- returns the results as a slice of Values, one per formal result.

Can you add some comment why "nil" is fine here or just explain it to me. Thanks.

Reply:
`nil` is a valid zero-length slice.